libs/wingc/src/lsp/mod.rs

```diff
@@ -0,0 +1,289 @@
+use crate::{
+	diagnostic::DiagnosticLevel,
+	parser::Parser,
+	type_check,
+	wasm_util::{combine_ptr_and_length, ptr_to_string},
+	Diagnostics,
+};
+use itertools::Itertools;
+use lsp_types::{
+	CompletionItem, CompletionResponse, Position, Range, SemanticToken, SemanticTokens, SemanticTokensResult, Url,
+};
+use std::{
+	cell::RefCell,
+	collections::{HashMap, HashSet},
+};
+use tree_sitter::Tree;
+
+use self::{completions::completions_from_ast, semantic_token::semantic_token_from_ast};
+
+mod completions;
+mod semantic_token;
+
+pub struct FileData {
+	pub contents: String,
+	pub tree: Tree,
+	pub diagnostics: Diagnostics,
+}
+
+thread_local! {
+	/// Under normal cases, wingc is not in full control of the process in which it is running.
```

Comment:
What does this mean ("under normal cases")?

Reply:
My intent was "When wingc is consumed as a WASM library", so I suppose I could write that instead 😁